File: robot_sdk/python/examples/test_robot_mqtt.py
#!/usr/bin/env python3
"""Example test robot using the Robomesh SDK over MQTT.

Demonstrates the full MQTT lifecycle:
1. Generate keys (or use existing ones)
2. Provision via admin API
3. Authenticate via MQTT topic-based challenge-response
4. Send heartbeats over MQTT
5. Exchange messages over MQTT

Requires: pip install paho-mqtt>=2.0

Usage:
    python test_robot_mqtt.py

    # Custom server:
    ROBOMESH_HOST=192.168.1.50 ROBOMESH_MQTT_PORT=1883 python test_robot_mqtt.py
"""

import logging
import os
import signal
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from robomesh_sdk import generate_ed25519_keypair
from robomesh_sdk.mqtt_client import RobotMQTTClient
from robomesh_sdk.admin import AdminClient

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(name)s] %(message)s")
logger = logging.getLogger("test_robot_mqtt")

HOST = os.environ.get("ROBOMESH_HOST", "localhost")
MQTT_PORT = int(os.environ.get("ROBOMESH_MQTT_PORT", "1883"))
HTTP_PORT = int(os.environ.get("ROBOMESH_HTTP_PORT", "8080"))


def main():
    # 1. Generate a fresh keypair
    _, priv_hex, pub_hex = generate_ed25519_keypair()
    robot_uuid = "test-robot-mqtt-demo"
    device_type = "test_robot"

    logger.info("Robot UUID:    %s", robot_uuid)
    logger.info("Public key:    %s", pub_hex)

    # 2. Provision via admin API
    admin = AdminClient(host=HOST, http_port=HTTP_PORT)
    admin.login("admin", "password1")
    try:
        admin.provision_robot(robot_uuid, pub_hex, device_type)
        logger.info("Robot provisioned successfully")
    except Exception as e:
        logger.warning("Provision failed (may already exist): %s", e)

    # 3. Connect and authenticate over MQTT
    client = RobotMQTTClient(
        uuid=robot_uuid,
        private_key_hex=priv_hex,
        host=HOST,
        mqtt_port=MQTT_PORT,
    )

    client.connect()
    jwt = client.authenticate()
    logger.info("Authenticated via MQTT! JWT: %s...%s", jwt[:20], jwt[-10:])

    # 4. Register message callback
    def on_message(msg):
        logger.info("Received from handler: %s", msg)

    client.on_message(on_message)

    # 5. Send heartbeats in background
    client.start_heartbeat(interval=25, ttl=60)

    # 6. Send some messages
    client.send("hello from MQTT test robot")
    time.sleep(1)
    client.send('{"type": "status", "battery": 88, "protocol": "mqtt"}')

    # 7. Wait for Ctrl+C
    logger.info("MQTT robot running. Press Ctrl+C to stop.")

    def handle_sigint(sig, frame):
        logger.info("Shutting down...")
        client.stop_heartbeat()
        client.disconnect()
        sys.exit(0)

    signal.signal(signal.SIGINT, handle_sigint)

    while True:
        time.sleep(1)


if __name__ == "__main__":
    main()

File: robot_sdk/c/examples/test_robot_mqtt.c
/**
 * Example test robot using the Robomesh C SDK over MQTT.
 *
 * Demonstrates: connect to MQTT broker → authenticate → heartbeat → messaging.
 * Uses the pre-seeded example-001 robot.
 * Requires libmosquitto.
 *
 * Usage:
 *   ./test_robot_mqtt [host] [mqtt_port]
 *   Default: localhost:1883
 */

#include "robomesh_mqtt.h"
#include "robomesh.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>

static volatile int running = 1;

static void sigint_handler(int sig) {
    (void)sig;
    running = 0;
}

static void on_message(const char *message, void *user_data) {
    (void)user_data;
    printf("Received from handler: %s\n", message);
}

int main(int argc, char *argv[]) {
    const char *host = argc > 1 ? argv[1] : "localhost";
    int mqtt_port = argc > 2 ? atoi(argv[2]) : 1883;

    /* Load the pre-seeded test robot's key */
    robomesh_keypair_t kp;
    robomesh_load_keypair("c55608b70c4a9f3b43bd1d23e86aaf4c3b2f4b823f54dc34ac668e85363ef2e1",
                           &kp);

    char pub_hex[65];
    robomesh_public_key_hex(&kp, pub_hex);
    printf("Robot: example-001\n");
    printf("Public key: %s\n", pub_hex);
    printf("Protocol: MQTT on %s:%d\n\n", host, mqtt_port);

    /* Create MQTT client */
    robomesh_mqtt_config_t config = {
        .uuid = "example-001",
        .host = host,
        .mqtt_port = mqtt_port,
        .keypair = &kp,
    };
    robomesh_mqtt_client_t *client = robomesh_mqtt_create(&config);
    if (!client) {
        fprintf(stderr, "Failed to create MQTT client\n");
        return 1;
    }

    /* Connect to broker */
    printf("Connecting to MQTT broker...\n");
    robomesh_err_t err = robomesh_mqtt_connect(client);
    if (err != ROBOMESH_OK) {
        fprintf(stderr, "MQTT connect failed: %s\n", robomesh_mqtt_last_error(client));
        robomesh_mqtt_destroy(client);
        return 1;
    }
    printf("Connected to MQTT broker\n");

    /* Register message callback */
    robomesh_mqtt_on_message(client, on_message, NULL);

    /* Authenticate */
    printf("Authenticating over MQTT...\n");
    err = robomesh_mqtt_authenticate(client, 10000);
    if (err != ROBOMESH_OK) {
        fprintf(stderr, "MQTT auth failed: %s\n", robomesh_mqtt_last_error(client));
        robomesh_mqtt_destroy(client);
        return 1;
    }
    const char *jwt = robomesh_mqtt_get_jwt(client);
    printf("Authenticated! JWT: %.20s...%s\n", jwt, jwt + strlen(jwt) - 10);

    /* Send a test message */
    printf("Sending test message...\n");
    err = robomesh_mqtt_send(client, "hello from MQTT test robot");
    if (err != ROBOMESH_OK) {
        fprintf(stderr, "Send failed: %s\n", robomesh_mqtt_last_error(client));
    } else {
        printf("Message sent OK\n");
    }

    /* Send heartbeats */
    signal(SIGINT, sigint_handler);
    printf("Sending heartbeats over MQTT. Press Ctrl+C to stop.\n");

    while (running) {
        err = robomesh_mqtt_send_heartbeat(client, NULL, 60);
        if (err != ROBOMESH_OK) {
            fprintf(stderr, "Heartbeat error: %s\n", robomesh_mqtt_last_error(client));
            break;
        }
        printf("MQTT heartbeat OK\n");
        sleep(25);
    }

    printf("Shutting down...\n");
    robomesh_mqtt_stop_heartbeat(client);
    robomesh_mqtt_destroy(client);
    return 0;
}
